compiler/fir/fir2ir/src/org/jetbrains/kotlin/fir/backend/Fir2IrVisitor.kt

```diff
@@ -760,10 +758,20 @@ class Fir2IrVisitor(
 
     override fun visitArrayOfCall(arrayOfCall: FirArrayOfCall, data: Any?): IrElement {
         return arrayOfCall.convertWithOffsets { startOffset, endOffset ->
+            lateinit var elementType: IrType
+            lateinit var arrayType: IrType
+            if (arrayOfCall.typeRef is FirResolvedTypeRef) {
+                arrayType = arrayOfCall.typeRef.toIrType()
+                elementType = arrayType.getArrayElementType(irBuiltIns)
+            } else {
+                // TODO: this should be the least upper bound of all arguments' types, e.g., ["4", 2u, 0.42f]
```

Comment:
Since this can only happen for annotation arguments, would it be possible to infer the correct type from the context?

Reply:
Perhaps I should have updated the (high-level) comment about this PR (will do so). As TODO above mentioned, this `else` case is for collection literals (according to other commit messages, or array literals), where we have not set type of the resulting array. So, we should infer the array type from the elements, and I believe it should be the least upper bound of all elements, not the first one. Note that this part was originally there, i.e., out of scope of this PR (I can revisit this later), and I guess, some cases like I mentioned, e.g., ["4", 2] which should have `Array<Any>`, would (silently) fail at this moment. :)

On the other hand, above `if` case corresponds to resolved `arrayOf` call where we know the return type of the call, which is now the array type, and then we rather infer the element type from it. This will be part of an answer to your other question (regarding `Array<Int>` v.s. `IntArray`), but I'll repeat anyway: once the call `arrayOf` is resolved and its return type is resolved as `Array<Int>`, `getArrayElementType` will return `Int` from that boxed array type, so no worries about unexpected conversion.

Reply:
I found `IrTypeUtils` has `commonSuperclass`, which is a good fit here.